src/circuit/gadget/ecc/chip/mul_fixed.rs

```diff
@@ -0,0 +1,580 @@
+        assert_eq!(
+            config.x_p, config.add_config.x_p,
+            "add is used internally in mul_fixed."
+        );
+        assert_eq!(
+            config.y_p, config.add_config.y_p,
+            "add is used internally in mul_fixed."
+        );
+
+        // Check relationships between this config and `add_incomplete_config`.
+        assert_eq!(
+            config.x_p, config.add_incomplete_config.x_p,
+            "add_incomplete is used internally in mul_fixed."
+        );
+        assert_eq!(
+            config.y_p, config.add_incomplete_config.y_p,
+            "add_incomplete is used internally in mul_fixed."
+        );
+        for advice in [config.x_p, config.y_p, config.window, config.u].iter() {
+            assert_ne!(
+                *advice, config.add_config.x_qr,
+                "Do not overlap with output columns of add."
+            );
+            assert_ne!(
+                *advice, config.add_config.y_qr,
+                "Do not overlap with output columns of add."
+            );
+        }
+
+        config
+    }
+}
+
+impl<const NUM_WINDOWS: usize> Config<NUM_WINDOWS> {
+    pub(super) fn create_gate_scalar(&self, meta: &mut ConstraintSystem<pallas::Base>) {
+        meta.create_gate(
+            "x_p, y_p checks for ScalarFixed, ScalarFixedShort",
+            |meta| {
+                let mul_fixed = meta.query_fixed(self.mul_fixed, Rotation::cur());
+                let window = meta.query_advice(self.window, Rotation::cur());
+                self.coords_check(meta, mul_fixed, window)
+            },
+        )
+    }
+
+    #[allow(clippy::op_ref)]
+    fn coords_check(
+        &self,
+        meta: &mut VirtualCells<'_, pallas::Base>,
+        toggle: Expression<pallas::Base>,
+        window: Expression<pallas::Base>,
+    ) -> Vec<Expression<pallas::Base>> {
+        let y_p = meta.query_advice(self.y_p, Rotation::cur());
+        let x_p = meta.query_advice(self.x_p, Rotation::cur());
+        let z = meta.query_fixed(self.fixed_z, Rotation::cur());
+        let u = meta.query_advice(self.u, Rotation::cur());
+
+        let window_pow: Vec<Expression<pallas::Base>> = (0..constants::H)
+            .map(|pow| {
+                (0..pow).fold(Expression::Constant(pallas::Base::one()), |acc, _| {
+                    acc * window.clone()
+                })
+            })
+            .collect();
+
+        let interpolated_x = window_pow.iter().zip(self.lagrange_coeffs.iter()).fold(
+            Expression::Constant(pallas::Base::zero()),
+            |acc, (window_pow, coeff)| {
+                acc + (window_pow.clone() * meta.query_fixed(*coeff, Rotation::cur()))
+            },
+        );
+
+        // Check interpolation of x-coordinate
+        let x_check = interpolated_x - x_p;
+        // Check that `y + z = u^2`, where `z` is fixed and `u`, `y` are witnessed
+        let y_check = u.clone() * u - y_p - z;
+
+        vec![toggle.clone() * x_check, toggle * y_check]
```

Comment:
Blocking: Unless it's somewhere else, this is missing the on-curve check:
```suggestion
        // Check interpolation of x-coordinate
        let x_check = interpolated_x - x_p.clone();
        // Check that `y + z = u^2`, where `z` is fixed and `u`, `y` are witnessed
        let y_check = u.square() - y_p.clone() - z;
        // Check that (x, y) is on the curve
        let on_curve = y_p.square() - x_p.clone().square() * x_p - 5;

        std::array::IntoIter::new([
            ("check x", x_check),
            ("check y", y_check),
            ("on-curve", on_curve),
        ]).map(move |(name, poly)| (name, toggle.clone() * poly))
```
(I don't remember if there's a constant for `b` in `pasta_curves`.)